### ISSUE TITLE
Remote connection to download an area does it multiple times.

### ISSUE THREAD

**Issue by RM87** [openstreetmap/merkaartor#147]
Testing:
enable remote control in merkaartor and download a random small area from openstreetmap.org webpage(Edit->Edit with remote control).

Fun part:

- With qt4 this seems to work ok (area is downloaded only once)
- With qt 5.7.1 and firefox 10 sockets are made (and area is downloaded 10 times) . This seems to be constant each time.
- With qt 5.7.1 and chrome only 2-3 sockets are made. 
- (running with qt 5.7.1 and having compiled with qt 5.7.0  + using firefox caused probably 0 sockets or something similar) 

Firefox shows that only one request was sent to localhost as expected.

In `MainWindow::incomingLocalConnection() `
new sockets are made multiple times here: 
`QTcpSocket *clientConnection = p->theListeningServer->nextPendingConnection();  `

If this gets figured out then in 
 `MainWindow::readLocalConnection()`
`resultStream << "OK\r\n"; `    can  be changed to  `resultStream << "OK\r\n"<< flush;` to fix osm website not receiving response from localhost. This can't be used currently due to segfault deep in qt code which probably comes from the above socket fun.

**Comment by Krakonos** [comment]
Interesting. The nextPendingConnection() call indeed returns multiple times with a different socket each time. I tried it with wget and it seems wget is retrying really fast. Probably because the disconnectFromHost() call. I think firefox is doing the same in the background. I will check with wireshark later, but the solution is probably to clean up the code a bit and make it send the response properly.

**Comment by Krakonos** [comment]
Hi! I did the cleanup and the code is in PR #156 for review and testing. It's in a state that "works for me" and handles the connection lifecycle much more transparently. With this patch, I only get one load and no error from the UI. Could you please verify the results?

I have found one additional issue where Merkaartor would crash after the load, especially when the window was not visible. This problem tracks down to modifying the Document object when the renderer is running. I have a quick & dirty fix for that, but unfortunately I am no longer able to reproduce the problem on top of this branch. I'd appreciate if you could check this as well:

1) Minimize Merkaartor
2) Use remote control to load an area
3) The map should download, but Merkaartor crashes at the end of "Parsing XML" dialog (or very soon after that).

**Comment by RM87** [comment]
1. Firefox 52.9 ESR has a bug with mixed content loading that simply kills remote control on osm website. 

Tested and it seems to work approx 25% of time. the rest segfaults. Being minimized or visible does not change the outcome.

**Comment by Krakonos** [comment]
Interesting. Do you mean there is a bug in firefox that causes the issues? By the stack trace I'd guess the fault is in our code and/or Qt.

**Comment by RM87** [comment]
Firefox is another issue. it is just buggy.

Tested with chrome. Merkaartor seems to crash while downloading new area.

**Comment by Krakonos** [comment]
Hi!
I just re-checked and I can't reproduce any problems on PR #156 . Is it related to the issue I mentioned in https://github.com/openstreetmap/merkaartor/issues/147#issuecomment-413766335 ? Can you verify if you are testing on that branch and if so, provide more details? Especially the Qt version.

Thanks!

**Comment by Krakonos** [comment]
Thanks for testing!

That might not be completely related. If you manage to reproduce it (or catch a stacktrace/sanitizer report), let me know. I will do some mapping this weekend to test it, and merge it if I don't find any outstanding issues.

**Comment by Krakonos** [comment]
As I have found no additional issues and none were reported, I merged PR #156 and consider this task done. If anything else shows up, please create a new issue.
